src/qmozwindow.h

```diff
@@ -24,6 +24,9 @@ class QMozWindow: public QObject
     explicit QMozWindow(const QSize &size, QObject *parent = nullptr);
     ~QMozWindow();
 
+    void createWindow();
+    void destroyWindow();
```

Comment:
This is leaking internals. The QMozWindow is already the thing and have public methods that can be used to destroy internals go to wrong direction.

Reply:
A separate createWindow() is needed to start window creation after all QMozWindow signals are connected. It is used here:: https://github.com/sailfishos/sailfish-browser/pull/897/commits/6f8e14e99bdbefbad5eee0bc60a13f0ead496359 .
EmbedLiteWindowBaseChild (compositor thread) calls QMozWindowPrivate methods directly. To destroy QMozWindow, we must be sure that the corresponding EmbedLiteWindowBaseChild is destroyed. Otherwise, the EmbedLiteWindowBaseChild may call DrawUnderlay or RequestGLContext methods of the destroyed QMozWindow. To prevent this undefined behavior, two steps are needed: send destroy to the EmbedLiteWindowBaseChild object, and when it is destroyed, only then delete QMozWindow object. The only way to get rid of destroyWindow() I see is to somehow get rid of using EmbedLiteWindowListener on the Child side (GLContext can be passed when the window surface is created and DrawUnderlay and DrawOverlay can be implemented on the embedlite side somehow, no idea how yet). Correct me please if I am wrong.

Reply:
You missed the point. This introduces to the two  public methods that reflect constructor and destructor but are not the same. I'd try to avoid leaking create/destroy kind of methods from the interface.

Possibility of calling DrawUnderlay and RequestGLContext of destroyed QMozWindow is indeed problem. Simple solution there could be that EmbedLiteWindow::Destroy() would nullify mListener. After this all calls to listener would be going to fake listener instead of the QMozWindow instance.

Reply:
I agree and also want to avoid leaking create/destroy kind of methods from the interface. But actually, I see no way to do this right now.
Nullifing mListener is not a safe approach, since getting a pointer and calling a method by it is not an atomic operation. 
Compositor thread may hang between GetListener() and RequestGLContext because of system scheduler decision 
https://github.com/sailfishos/gecko-dev/blob/b9e7f6a3415c2473b0eb39496a1e13c36d68cd14/embedding/embedlite/embedshared/nsWindow.cpp#L380
Of course this is an ultra rare case, but should we rely on that...
I also removed EmbedLiteWindowListener::mListener and Set/GetListener because it only works if the order of the calls is as expected:
1. https://github.com/sailfishos/qtmozembed/blob/b71492b065c0d879dc67d756d6d2f081aaf16372/src/qmozwindow.cpp#L28
2. https://github.com/sailfishos/gecko-dev/blob/b9e7f6a3415c2473b0eb39496a1e13c36d68cd14/embedding/embedlite/EmbedLiteApp.cpp#L484
3. https://github.com/sailfishos/qtmozembed/blob/b71492b065c0d879dc67d756d6d2f081aaf16372/src/qmozwindow.cpp#L29
4. https://github.com/sailfishos/gecko-dev/blob/b9e7f6a3415c2473b0eb39496a1e13c36d68cd14/embedding/embedlite/embedshared/nsWindow.cpp#L376
But very often 4 happens before 2 or 3. And then GetListener() returns a null pointer and the browser draws a white screen. The only stable option I've found is to pass the Listener directly to the EmbedLiteWindowBaseChild constructor. (this is problem #2)
If we want to nullify the pointer in the EmbedLiteViewBaseChild, we need direct access from the EmbedLiteApp. Didn't check if we have it. If not, I need return EmbedLiteWindowListener::mListener and Set/GetListener back and look for another solution for problem #2.

Reply:
Thanks for the update. We could try to mask this a bit with method names
```
reserve();
release();
isReserved();
```
Maybe something like above? What do you think? isReserved() would imply that mozwindow is valid => should the method be called isValid() actually? @llewelld any thoughts?